### PR TITLE
Add support for delay values split into parts + millisecond time input

### LINKED ISSUE #4733: When setting Delay with yaml params, the interface sets the delay to '[object Object]'
## The problem
When setting a delay (as in the eg below) in a script / automation using the yaml interface, then re-editing with the interface, the delay in the yaml becomes '[object Object]' (an invalid delay)...

## Environment

- Home Assistant release with the issue: 0.104.3
- Last working Home Assistant release (if known): 
- Operating environment (Hass.io/Docker/Windows/etc.): Docker
- Integration causing this issue: Script Syntax
- Link to integration documentation on our website: https://www.home-assistant.io/docs/scripts/

## Problem-relevant `configuration.yaml`

```yaml
delay:
  milliseconds: 250
#When re-saving this is changed to:
delay: '[object Object]'
```

### LINKED ISSUE #5542: Automation/Script Editor break delay
## The problem
I moved my scripts into scripts.yaml to use script editor, same for automations. If I now try to edit the script/automation the delay data got loaded as [object Object] so I am not able to edit it. And saving break the script/automation because it also get written into files as [object Object]. 

## Environment
Home Assistant 0.108.3

- Home Assistant Core release with the issue: 0.108.3
- Last working Home Assistant Core release (if known): Never
- Operating environment (Home Assistant/Supervised/Docker/venv): Docker
- Integration causing this issue: 
- Link to integration documentation on our website: 

## Problem-relevant `scripts.yaml`

```yaml
bedroom_underground_light_off:
  alias: Schalte Untergrund Bett Beleuchtung nach 30 Sekunden aus
  sequence:
  - delay:
      seconds: 30
  - service: switch.turn_off
    data:
      entity_id: switch.bedroom_bed_light_switch

```

## Traceback/Error logs

```txt

```

### LINKED ISSUE #6503: Backspacing duration in delay action results in Unsupported Action
## Checklist

- [ YES ] I have updated to the latest available Home Assistant version.
- [ YES ] I have cleared the cache of my browser.
- [ YES ] I have tried a different browser to see if it is related to my browser.

## The problem

Creating an automation through the UI, when you add a delay action and have entered a duration, backspacing the entered duration loses the action and sets it to an unsupported action

## Expected behavior

The UI should allow you to backspace the duration and when empty not take any future action

## Steps to reproduce

- create a new automation and complete a trigger (any will do)
- add an action (say turn device on)
- add a secondary action of delay
-- within the delay second field (watermarked Delay) enter a delay duration (for example 00:01 - highlighted in screen shot 1 in red)
- you can save if required or continue the steps to replicate
- with the cursor in the field with the duration (highlighted in red in screen shot 1), backspace the entered value out, once you have reached the final character (in our case 0), backspace that out
- the UI then deletes the action of delay and sets you to an unsuported action (screen shot 2)

image

image

## Environment

- Home Assistant release with the issue: 0.113
- Last working Home Assistant release (if known): unknown
- Browser and browser version: Chrome Version 84.0.4147.105 (tested in edge as well)
- Operating system: HassOS 4.11

## State of relevant entities

N/A 

## Problem-relevant configuration

N/A - easily replicated without prior configuration

### PR DESCRIPTION
**Pull request by spacegaier**
## Proposed change

Adds support for handling both string values ("01:02:03:456") as well as delay values split into parts (hours, minutes, etc.).
Both YAML versions are now shown via a time picker that has been extended to also support milliseconds. Days are handled by converting them to hours.

image

## Type of change

- [ ] Dependency upgrade
- [X] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes https://github.com/home-assistant/frontend/issues/4733 fixes https://github.com/home-assistant/frontend/issues/5542 fixes https://github.com/home-assistant/frontend/issues/6503
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [X] The code change is tested and works locally.
- [X] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/components/paper-time-input.js [hunk 1]
```diff
@@ -167,6 +171,28 @@ export class PaperTimeInput extends PolymerElement {
           always-float-label$="[[alwaysFloatInputLabels]]"
           disabled="[[disabled]]"
           hidden$="[[!enableSecond]]"
+        >
+          <span hidden$="[[!enableMillisecond]]" suffix slot="suffix">:</span>
+        </paper-input>
+
+        <!-- Millisec Input -->
+        <paper-input
```
Comment (bramkragten) [inline src/components/paper-time-input.js:179]: We should really make this element a LitElement, it is so complicated now...

File: src/panels/config/automation/action/types/ha-automation-action-delay.ts [hunk 2]
```diff
@@ -17,22 +22,53 @@ export class HaDelayAction extends LitElement implements ActionElement {
   }
 
   protected render() {
-    const { delay } = this.action;
+    let data: HaFormTimeData = {};
+
+    if (typeof this.action.delay !== "object") {
+      const parts = this.action.delay?.toString().split(":") || [];
+      data = {
+        hours: +parts[0],
```
Comment (bramkragten) [inline src/panels/config/automation/action/types/ha-automation-action-delay.ts:30]:
```suggestion
        hours: Number(parts[0]),
```

File: src/panels/config/automation/action/types/ha-automation-action-delay.ts [hunk 3]
```diff
@@ -17,22 +22,53 @@ export class HaDelayAction extends LitElement implements ActionElement {
...
+    const schema = {} as HaFormTimeSchema;
 
     return html`
-      <paper-input
-        .label=${this.hass.localize(
-          "ui.panel.config.automation.editor.actions.type.delay.delay"
-        )}
-        name="delay"
-        .value=${delay}
+      <ha-form-positive_time_period_dict
```
Comment (bramkragten) [inline src/panels/config/automation/action/types/ha-automation-action-delay.ts:50]: ha-form elements should only be used inside ha-form

File: src/components/paper-time-input.js (hunk 1 above)
Comment (spacegaier) [reply src/components/paper-time-input.js:179]: @bramkragten That is not something I can manage before the beta and I think this fix should be in. I tackled all your other remarks.

Comment (bramkragten) [reply src/components/paper-time-input.js:179]: Should not be in this PR anyway

File: src/panels/config/automation/action/types/ha-automation-action-delay.ts [hunk 4]
```diff
@@ -17,22 +22,56 @@ export class HaDelayAction extends LitElement implements ActionElement {
...
+    };
 
     return html`
-      <paper-input
-        .label=${this.hass.localize(
-          "ui.panel.config.automation.editor.actions.type.delay.delay"
-        )}
-        name="delay"
-        .value=${delay}
+      <ha-form
```
Comment (bramkragten) [inline src/panels/config/automation/action/types/ha-automation-action-delay.ts:54]: We should not use `ha-form` here, you can extract what you need and create a separate component that you share between here and `ha-form`

File: src/panels/config/automation/action/types/ha-automation-action-delay.ts [hunk 5]
```diff
@@ -17,22 +19,49 @@ export class HaDelayAction extends LitElement implements ActionElement {
...
+    const configValue = "delay";
+    const value = ev.detail.value;
+
+    const newValue = { ...this.action };
+    if (value) {
+      newValue[configValue] = ev.detail.value;
+    }
+    fireEvent(this, "value-changed", {
+      value: newValue,
+    });
```
Comment (bramkragten) [inline src/panels/config/automation/action/types/ha-automation-action-delay.ts:64]:
```suggestion
  private _valueChanged(ev: CustomEvent) {
    ev.stopPropagation();
    const value = ev.detail.value;
	if (!value) {
		return;
	}
    fireEvent(this, "value-changed", {
      value: { ...this.action, delay: value },
    });
```

**Review by bramkragten — APPROVED** [review]

**Pull request merged**